### ISSUE TITLE
Slow typing in large file in large project

### ISSUE THREAD

**Issue by DutchEllie** [cyrusfirheir/twee3-language-tools#79]
Hi, when I use this extension to help me with writing it shows significant slowdowns when typing in a large file. The file is approximately 1000 lines long, but the slowdown occurs at a couple hundred lines too. This happens specifically in a very large project, with about 7700 passages. When all the passages are loaded, the performance is horrible, but when the passages are still loading and only about 100 are loaded, the performance is great. Therefore I think it has something to do with the amount of loaded passages. Is there anything that can be done about this?

I use Linux kernel 5.15.13-arch1-1 using Arch Linux.  
I run KDE Plasma 5.23.5 on X11 (not sure if that's important at all).  
I run VSCode version 1.63.2 from the [visual-studio-code-bin](https://aur.archlinux.org/packages/visual-studio-code-bin) package from the AUR.  
This is happening on a Lenovo ThinkPad L15 G2 AMD, which has a Ryzen 7 Pro 5850U and 16GB of RAM. 

I don't know what more information to provide right now. Do not hesitate to ask me for any more information. Thanks for developing this awesome project!

**Comment by Tranberry** [comment]
curious; does this happen even when you do not have passage listing turned ON? 

![image](https://user-images.githubusercontent.com/16529/148825945-e6c2bc82-028e-4bb6-a051-e864ba1268ce.png)

**Comment by DutchEllie** [comment]
I have not disabled that in the `settings.json` but I did try to disable it once to see if the problem would go away, it didn't.

**Comment by MinusGix** [comment]
The best thing to do to isolate it is by turning off certain settings and then reloading vscode (Ctrl + P -> Reload can make it faster than completely turning it off and on), try:  
- Parameter Validation (This does a variety of type checking to try to ensure that macros are being passed the right types)
- then try also disabling, Argument Parsing (Does a variety of parsing)
- then maybe try disabling Undefined Macro
It might be one of those, since they're some of the more expensive operations. Though, they haven't been optimized anywhere near as much as they could be, so if I can isolate it then that can help in improving things.  
I should probably generate a large project with random macros + lorem ipsum to make it easier to notice and find performance issues, since there isn't an open source ones for me to test with.

**Comment by DutchEllie** [comment]
I will try the suggestions you made and report back as soon as I can.

**Comment by DutchEllie** [comment]
I got it! The issue disappeared entirely when I set `twee3LanguageTools.sugarcube-2.error.argumentParsing` to `false`. For sanity, when I set it back to `true` the problem returns and setting it to `false` again fixes it immediately!

**Comment by MinusGix** [comment]
> but I am slightly anxious to share it publicly. It's not exactly the kind of project I need to be associated with...  
  
Yeah I get that, but that actually gave me an idea of an existing project to try testing that I forgot about (it might even be the one you're referring to, or might not). Hopefully, I'll be able to look at that to see if there's a part of the argument parsing that I can speed up notably.
